### PR TITLE
C++: Support non-type template parameters

### PR DESCRIPTION
**Pull request by jketema**
#### All query authors

- [x] A change note is added if necessary. See [the documentation](https://github.com/github/codeql/blob/main/docs/change-notes.md) in this repository.
- [ ] All new queries have appropriate `.qhelp`. See [the documentation](https://github.com/github/codeql/blob/main/docs/query-help-style-guide.md) in this repository.
- [ ] QL tests are added if necessary. See [Testing custom queries](https://docs.github.com/en/code-security/codeql-cli/using-the-advanced-functionality-of-the-codeql-cli/testing-custom-queries) in the GitHub documentation.
- [ ] New and changed queries have correct query metadata. See [the documentation](https://github.com/github/codeql/blob/main/docs/query-metadata-style-guide.md) in this repository.

#### Internal query authors only

- [ ] Autofixes generated based on these changes are valid, only needed if this PR makes significant changes to `.ql`, `.qll`, or `.qhelp` files. See [the documentation](https://github.com/github/codeql-team/blob/main/docs/best-practices/validating-autofix-for-query-changes.md) (internal access required).
- [x] Changes are validated [at scale](https://github.com/github/codeql-dca/) (internal access required).
- [ ] Adding a new query? Consider also [adding the query to autofix](https://github.com/github/codeml-autofix/blob/main/docs/updating-query-support.md#adding-a-new-query-to-the-query-suite).

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/TemplateParameter.qll [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+/**
+ * A C++ non-type template parameter.
+ *
+ * In the example below, `I` is a non-type template parameter:
+ * ```
+ * template <int I>
+ * class C { };
+ * ```
+ */
+class NonTypeTemplateParameter extends Literal, TemplateParameterImpl {
```
Comment (calumgrant) [inline cpp/ql/lib/semmle/code/cpp/TemplateParameter.qll:32]: I would rename this to `LiteralTemplateParameter`. If there are other types of non-type template parameter we can add them.

File: cpp/ql/lib/semmle/code/cpp/TemplateParameter.qll [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  override string getAPrimaryQlClass() { result = "TemplateTemplateParameter" }
+}
+
+/**
+ * A type representing the use of the C++11 `auto` keyword.
+ * ```
+ * auto val = some_typed_expr();
+ * ```
+ */
+class AutoType extends TypeTemplateParameter {
```
Comment (calumgrant) [inline cpp/ql/lib/semmle/code/cpp/TemplateParameter.qll:86]: This was the old name, but it's not obvious to me that `AutoType` is only for template parameters.

File: cpp/ql/lib/semmle/code/cpp/TemplateParameter.qll (hunk 1 above)
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/TemplateParameter.qll:32]: "non-type" is the name used by the C++ standard, so I think "NonTypeTemplateParameter" is most appropriate.

File: cpp/ql/lib/semmle/code/cpp/TemplateParameter.qll (hunk 2 above)
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/TemplateParameter.qll:86]: It's not (see also the example in the QLDoc), but I prefer not messing with this at this point.

File: cpp/ql/lib/semmle/code/cpp/TemplateParameter.qll (hunk 1 above)
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/TemplateParameter.qll:32]: It's a literal sort of by accident, due to the way our frontend chooses to represent these.

**Review by calumgrant — APPROVED** [review]

**Pull request merged**